controllers/release/release_adapter_test.go

```diff
@@ -0,0 +1,414 @@
+		applicationSnapshot.TypeMeta.Kind = "applicationSnapshot"
+		Expect(noCacheClient.Update(ctx, applicationSnapshot)).Should(Succeed())
+		pipelineRun, err := adapter.createReleasePipelineRun(
+			releaseStrategy,
+			applicationSnapshot)
+		Expect(pipelineRun != nil && err == nil).To(BeTrue())
+		Expect(noCacheClient.Delete(ctx, pipelineRun)).Should(Succeed())
+	})
+
+	It("can finalize (delete) the releasePipelineRun just processed", func() {
+		result, err := cacheAdapter.EnsureFinalizersAreCalled()
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
+
+		Eventually(func() bool {
+			result, err := cacheAdapter.EnsureFinalizerIsAdded()
+			return (!result.CancelRequest && err == nil)
+		}).Should(BeTrue())
+
+		Eventually(func() bool {
+			result, err := cacheAdapter.EnsureReleasePipelineRunExists()
+			return (!result.CancelRequest && err == nil)
+		}).Should(BeTrue())
+		result, err = cacheAdapter.EnsureReleasePipelineStatusIsTracked()
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
+
+		err = cacheAdapter.finalizeRelease()
+		Expect(err).Should(Succeed())
+
+		pipelineRun, _ := cacheAdapter.getReleasePipelineRun()
+		Expect(cacheClient.Delete(ctx, pipelineRun)).ShouldNot(Succeed())
+	})
+
+	It("can return the ApplicationSnapshot from the release on the given adapter", func() {
+		snapshot, err := adapter.getApplicationSnapshot()
+		Expect(err).Should(Succeed())
+		Expect(reflect.TypeOf(snapshot)).
+			To(Equal(reflect.TypeOf(&appstudioshared.ApplicationSnapshot{})))
+
+		// should error when applicationSnapshot is missing
+		Expect(noCacheClient.Delete(ctx, applicationSnapshot)).Should(Succeed())
+		_, err = adapter.getApplicationSnapshot()
+		Expect(err).Should(HaveOccurred())
+	})
+
+	It("can return the releaseLink from the release within the given adapter", func() {
+		link, err := adapter.getReleaseLink()
+		Expect(err).Should(Succeed())
+		Expect(reflect.TypeOf(link)).
+			To(Equal(reflect.TypeOf(&appstudiov1alpha1.ReleaseLink{})))
+
+		// shold err when releaseLink does not exist
+		noCacheClient.Delete(ctx, releaseLink)
+		_, err = adapter.getReleaseLink()
+		Expect(err).Should(HaveOccurred())
+	})
+
+	It("can return the pipelineRun from the release within the give adapter", func() {
+		Eventually(func() bool {
+			result, err := cacheAdapter.EnsureReleasePipelineRunExists()
+			return (!result.CancelRequest && err == nil)
+		}).Should(BeTrue())
+		pipelineRun, err := cacheAdapter.getReleasePipelineRun()
+		Expect(err).Should(Succeed())
+		Expect(reflect.TypeOf(pipelineRun)).
+			To(Equal(reflect.TypeOf(&tektonv1beta1.PipelineRun{})))
+	})
+
+	It("can return the releaseStrategy from a given releaseLink to the given adapter", func() {
+		strategy, err := adapter.getReleaseStrategy(releaseLink)
+		Expect(err).Should(Succeed())
+		Expect(reflect.TypeOf(strategy)).
+			To(Equal(reflect.TypeOf(&appstudiov1alpha1.ReleaseStrategy{})))
+	})
+
+	It("can return the releaseStrategy from a given adapter", func() {
```

Comment:
Totally. Also, the name should be the one referenced in the ReleaseLink.

Reply:
Ok, added: 
```		Expect(strategy.Namespace).Should(ContainSubstring(link.Spec.Target))
```

Reply:
I think the added check is good, but looking again I actually think it makes more sense being in the test before `can return the releaseStrategy from a given releaseLink` since you already have the releaselink there. Can you just move this line to that test instead?

Reply:
Moved. Hope I did move the right one as I changed the description :D last week.